src/util/hooks.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+
+import * as React from 'react'
+
+type SetState<S> = (value: S | ((state: S) => S)) => void
+
+type UseCallback = <T>(callback: T, deps: any[]) => T
+
+type UseContext = <T>(context: React.Context<T>) => T
+
+type UseDebugValue = <T>(value: T, format?: (value: T) => any) => void
+
+type UseEffect = (effect: () => void | (() => void), deps?: any[]) => void
+
+type UseImperativeHandle = (ref: any, init: () => any, deps?: any[]) => void
+
+type UseMemo = <T>(init: () => T, deps?: any[]) => T
+
+type UseReducer = {
+  // Normal version:
+  <State, Action>(reducer: (state: State | void, action: Action) => State, init: State | void): [State, (action: Action) => void],
```

Comment:
Can you do something like this instead of duplicating the type definition?

`
<State, Action, Init = void>(
    reducer: (state: State | void, action: Action) => State,
    init: Init | State | void,
    initializer?: (init: Init) => State
  ): [State, (action: Action) => void]
  `

Reply:
No, because it's not actually correct. In the case with 2 arguments, the initial value must be `State`, but in the case with 3 values, it just has to match whatever argument 3 expects. Since argument 2 has different types depending on the argument count, we need two separate versions.

Reply:
That's why I was asking 👍🏼